include/fplus/benchmark_session.hpp

```diff
@@ -0,0 +1,359 @@
+//   (although it also works if it is not present)
+// - you can also benchmark an expression that returns void using benchmark_void_expression
+//
+template<class Fn>
+auto make_benchmark_function(benchmark_session & session, const FunctionName & name, Fn f)
+{
+    // transforms f into a function with the same 
+    // signature, that will store timings into the benchmark session
+    return internal::bench_function_impl<Fn>(session, name, f);
+}
+
+
+// API search type: make_benchmark_void_function : (benchmark_session, string, (a... -> Void)) -> (a... -> Void)
+// fwd bind count: 0
+// Transforms a function that returns a void into a function with the *same* signature
+// and behavior, except that it also stores stats into the benchmark session (first parameter),
+// under the name given by the second parameter
+//
+// Note that make_benchmark_void_function *will add side effects* to the function
+// (since it stores data into the benchmark session at each call)
+//
+// Example:
+//
+// benchmark_session bench_session;
+// ...
+// void foo() { 
+//     std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+// }
+// ...
+// auto foo_bench = make_benchmark_void_function(bench_session, "foo", foo);
+// foo_bench();
+// ...
+// std::cout << benchmark_session.report();
+template<class Fn>
+auto make_benchmark_void_function(benchmark_session & session, const FunctionName & name, Fn f)
+{
+    // transforms a void returning function into a function with the same 
+    // signature, that will store timings into the benchmark session
+    return internal::bench_void_function_impl<Fn>(session, name, f);
+}
+
+#define benchmark_expression(bench_session, name, expression)      \
+make_benchmark_function(                                           \
+    bench_session,                                                 \
+    name,                                                          \
+    [&]() { return expression; }                                   \
+)();
+
+#define benchmark_void_expression(bench_session, name, expression) \
+make_benchmark_void_function(                                      \
+    bench_session,                                                 \
+    name,                                                          \
+    [&]() { expression; }                                          \
+)();
+
+
+template<class Fn>
+auto run_n_times(int nb_runs, Fn f) // run_n_times is perhaps a nice addition to fplus
+{
+    for (auto _ : fplus::numbers(0, nb_runs)) {
+        (void) _; // suppress warning / unused variable
+        f();
+    }
+}
+
+
+namespace internal
+{
+    // transform_and_sum is perhaps a nice addition to fplus
+    template <typename UnaryF, typename Container>
+    auto transform_and_sum(UnaryF unary_f, const Container& xs)
+    {
+        return fplus::sum(fplus::transform(unary_f, xs));
+    }
+
+
+    inline std::string show_tableau(const std::vector<std::vector<std::string>> & rows)
```

Comment:
Bike shedding: Please no space before the `&`. :wink:

Question: How about the term "table" instead of "tableau"?

Reply:
done